INTLY-5489 CLI to check all namespaces when flag is not present

sergioifg94 opened the pull request:
> [Link to JIRA](https://issues.redhat.com/browse/INTLY-5489)

Add logic in CLI to get a report for all namespaces in the cluster when the `-namespaces` flag is not present in the command argument.

sergioifg94 committed:
feat: CLI to check all namespaces when flag is not present

Add logic in CLI to get a report for all namespaces in the cluster when
the `-namespaces` flag is not present in the command argument.

sergioifg94 commented:
/assign @Nanyte25

Nanyte25 commented:
@sergioifg94  I have tested your pr on our rhmi-cssre staging cluster. Once I rebuilt the go binary it worked for a time but bailed out on the with an error see below. A suggestion might be to search all openshift-* namespaces as it does really need to check all namespace on a given cluster.

```
./cli
2020/04/20 10:03:03 generating report for all namespaces
checking image : registry.access.redhat.com/rhscl/httpd-24-rhel7:latest
skipping tag  latest  as it does not match on major minor patch version 2.4.*
skipping tag  2.4  as it does not match on major minor patch version 2.4.*
skipping tag  2.4  as it does not match on major minor patch version 2.4.*
skipping tag  latest  as it does not match on major minor patch version 2.4.*
skipping tag  2.4  as it does not match on major minor patch version 2.4.*
skipping tag  latest  as it does not match on major minor patch version 2.4.*
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
| COMPONENT                     | IMAGE                | IMAGE HASH                                                       | IMAGE STREAM | TAG    | UPTO DATE WITH TAG | PERSISTENT IMAGE TAG | LATEST PATCH TAG | FLOATING TAG | USING FLOATING TAG | UPTO DATE WITH FLOATING TAG | CRITICAL CVES | IMPORTANT CVES | MODERATE CVES |
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
| web-console-config-web-server | rhscl/httpd-24-rhel7 | d726ce74164b1b757c5f6523c480ae0a84fdd09ae5ad8e2252a8e185c8f9273e | false        | latest | false              | 2.4-109              | 2.4-110          | latest       | true               | false                       |             0 |              0 |             0 |
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
| COMPONENT                     | IMAGE                | IMAGE HASH                                                       | IMAGE STREAM | TAG    | UPTO DATE WITH TAG | PERSISTENT IMAGE TAG | LATEST PATCH TAG | FLOATING TAG | USING FLOATING TAG | UPTO DATE WITH FLOATING TAG | CRITICAL CVES | IMPORTANT CVES | MODERATE CVES |
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
| web-console-config-web-server | rhscl/httpd-24-rhel7 | d726ce74164b1b757c5f6523c480ae0a84fdd09ae5ad8e2252a8e185c8f9273e | false        | latest | false              | 2.4-109              | 2.4-110          | latest       | true               | false                       |             0 |              0 |             0 |
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
| COMPONENT                     | IMAGE                | IMAGE HASH                                                       | IMAGE STREAM | TAG    | UPTO DATE WITH TAG | PERSISTENT IMAGE TAG | LATEST PATCH TAG | FLOATING TAG | USING FLOATING TAG | UPTO DATE WITH FLOATING TAG | CRITICAL CVES | IMPORTANT CVES | MODERATE CVES |
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
| web-console-config-web-server | rhscl/httpd-24-rhel7 | d726ce74164b1b757c5f6523c480ae0a84fdd09ae5ad8e2252a8e185c8f9273e | false        | latest | false              | 2.4-109              | 2.4-110          | latest       | true               | false                       |             0 |              0 |             0 |
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
| COMPONENT                     | IMAGE                | IMAGE HASH                                                       | IMAGE STREAM | TAG    | UPTO DATE WITH TAG | PERSISTENT IMAGE TAG | LATEST PATCH TAG | FLOATING TAG | USING FLOATING TAG | UPTO DATE WITH FLOATING TAG | CRITICAL CVES | IMPORTANT CVES | MODERATE CVES |
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
| web-console-config-web-server | rhscl/httpd-24-rhel7 | d726ce74164b1b757c5f6523c480ae0a84fdd09ae5ad8e2252a8e185c8f9273e | false        | latest | false              | 2.4-109              | 2.4-110          | latest       | true               | false                       |             0 |              0 |             0 |
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
panic: runtime error: index out of range [1] with length 1

goroutine 1 [running]:
github.com/integr8ly/heimdall/pkg/cluster.ParseImage(0xc0006e2280, 0x3e, 0xc0004145a0)
	/Users/markfreer/repos/heimdall/pkg/cluster/images.go:191 +0x5b0
github.com/integr8ly/heimdall/pkg/cluster.(*ImageService).FindImagesFromLabels(0xc0002dea40, 0xc000423399, 0x7, 0xc00062b0e0, 0x2, 0x2, 0x33, 0x0, 0x0)
	/Users/markfreer/repos/heimdall/pkg/cluster/images.go:149 +0xeec
github.com/integr8ly/heimdall/pkg/controller/deploymentconfigs.(*Reports).GetImages(0xc000545c00, 0xc0005453f0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/markfreer/repos/heimdall/pkg/controller/deploymentconfigs/report.go:109 +0x49e
github.com/integr8ly/heimdall/pkg/controller/deploymentconfigs.(*Reports).Generate(0xc000545c00, 0xc000404a84, 0x7, 0x1fd2299, 0x1, 0x0, 0x0, 0xc00072f930, 0x7, 0xc00072f870)
	/Users/markfreer/repos/heimdall/pkg/controller/deploymentconfigs/report.go:60 +0xb09
main.main()
	/Users/markfreer/repos/heimdall/cmd/cli/main.go:57 +0xc63

```
@sergioifg94  This is work for no input and the new option we suggest of all openshift-* perfix namespaces.


```

/cli
checking image : registry.access.redhat.com/rhscl/httpd-24-rhel7:latest
skipping tag  latest  as it does not match on major minor patch version 2.4.*
skipping tag  2.4  as it does not match on major minor patch version 2.4.*
skipping tag  2.4  as it does not match on major minor patch version 2.4.*
skipping tag  latest  as it does not match on major minor patch version 2.4.*
skipping tag  2.4  as it does not match on major minor patch version 2.4.*
skipping tag  latest  as it does not match on major minor patch version 2.4.*
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
| COMPONENT                     | IMAGE                | IMAGE HASH                                                       | IMAGE STREAM | TAG    | UPTO DATE WITH TAG | PERSISTENT IMAGE TAG | LATEST PATCH TAG | FLOATING TAG | USING FLOATING TAG | UPTO DATE WITH FLOATING TAG | CRITICAL CVES | IMPORTANT CVES | MODERATE CVES |
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
| web-console-config-web-server | rhscl/httpd-24-rhel7 | d726ce74164b1b757c5f6523c480ae0a84fdd09ae5ad8e2252a8e185c8f9273e | false        | latest | false              | 2.4-109              | 2.4-110          | latest       | true               | false                       |             0 |              0 |             0 |
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
| COMPONENT                     | IMAGE                | IMAGE HASH                                                       | IMAGE STREAM | TAG    | UPTO DATE WITH TAG | PERSISTENT IMAGE TAG | LATEST PATCH TAG | FLOATING TAG | USING FLOATING TAG | UPTO DATE WITH FLOATING TAG | CRITICAL CVES | IMPORTANT CVES | MODERATE CVES |
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
| web-console-config-web-server | rhscl/httpd-24-rhel7 | d726ce74164b1b757c5f6523c480ae0a84fdd09ae5ad8e2252a8e185c8f9273e | false        | latest | false              | 2.4-109              | 2.4-110          | latest       | true               | false                       |             0 |              0 |             0 |
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
| COMPONENT                     | IMAGE                | IMAGE HASH                                                       | IMAGE STREAM | TAG    | UPTO DATE WITH TAG | PERSISTENT IMAGE TAG | LATEST PATCH TAG | FLOATING TAG | USING FLOATING TAG | UPTO DATE WITH FLOATING TAG | CRITICAL CVES | IMPORTANT CVES | MODERATE CVES |
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
| web-console-config-web-server | rhscl/httpd-24-rhel7 | d726ce74164b1b757c5f6523c480ae0a84fdd09ae5ad8e2252a8e185c8f9273e | false        | latest | false              | 2.4-109              | 2.4-110          | latest       | true               | false                       |             0 |              0 |             0 |
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
| COMPONENT                     | IMAGE                | IMAGE HASH                                                       | IMAGE STREAM | TAG    | UPTO DATE WITH TAG | PERSISTENT IMAGE TAG | LATEST PATCH TAG | FLOATING TAG | USING FLOATING TAG | UPTO DATE WITH FLOATING TAG | CRITICAL CVES | IMPORTANT CVES | MODERATE CVES |
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
| web-console-config-web-server | rhscl/httpd-24-rhel7 | d726ce74164b1b757c5f6523c480ae0a84fdd09ae5ad8e2252a8e185c8f9273e | false        | latest | false              | 2.4-109              | 2.4-110          | latest       | true               | false                       |             0 |              0 |             0 |
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
| COMPONENT                     | IMAGE                | IMAGE HASH                                                       | IMAGE STREAM | TAG    | UPTO DATE WITH TAG | PERSISTENT IMAGE TAG | LATEST PATCH TAG | FLOATING TAG | USING FLOATING TAG | UPTO DATE WITH FLOATING TAG | CRITICAL CVES | IMPORTANT CVES | MODERATE CVES |
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
| web-console-config-web-server | rhscl/httpd-24-rhel7 | d726ce74164b1b757c5f6523c480ae0a84fdd09ae5ad8e2252a8e185c8f9273e | false        | latest | false              | 2.4-109              | 2.4-110          | latest       | true               | false                       |             0 |              0 |             0 |
+-------------------------------+----------------------+------------------------------------------------------------------+--------------+--------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+


```
The second option with the openshift-* prefix is also working  :

```
./cli -namespace-pattern openshift-.*
checking image : registry.redhat.io/3scale-amp2/apicast-gateway-rhel7:3scale2.7
checking image : registry.redhat.io/3scale-amp2/backend-rhel7:3scale2.7
checking image : registry.redhat.io/rhscl/redis-32-rhel7:3.2
skipping tag  latest  as it does not match on major minor patch version 3.2.*
checking image : registry.redhat.io/3scale-amp2/system-rhel7:3scale2.7
checking image : registry.redhat.io/3scale-amp2/memcached-rhel7:3scale2.7
checking image : registry.redhat.io/rhscl/mysql-57-rhel7:5.7
skipping tag  5.7  as it does not match on major minor patch version 5.7.*
skipping tag  latest  as it does not match on major minor patch version 5.7.*
checking image : registry.redhat.io/3scale-amp2/zync-rhel7:3scale2.7
checking image : registry.redhat.io/rhscl/postgresql-10-rhel7
+--------------------+-----------------------------------+------------------------------------------------------------------+--------------+-----------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
| COMPONENT          | IMAGE                             | IMAGE HASH                                                       | IMAGE STREAM | TAG       | UPTO DATE WITH TAG | PERSISTENT IMAGE TAG | LATEST PATCH TAG | FLOATING TAG | USING FLOATING TAG | UPTO DATE WITH FLOATING TAG | CRITICAL CVES | IMPORTANT CVES | MODERATE CVES |
+--------------------+-----------------------------------+------------------------------------------------------------------+--------------+-----------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
| apicast-production | 3scale-amp2/apicast-gateway-rhel7 | 26700c6cb5f43fa8061cee274caef5561646995a0e7d803273f394d9cdda1391 | true         | 3scale2.7 | true               | 1.16                 | 1.16             | 3scale2.7    | true               | true                        |             0 |              0 |             0 |
| apicast-staging    | 3scale-amp2/apicast-gateway-rhel7 | 26700c6cb5f43fa8061cee274caef5561646995a0e7d803273f394d9cdda1391 | true         | 3scale2.7 | true               | 1.16                 | 1.16             | 3scale2.7    | true               | true                        |             0 |              0 |             0 |
| backend-cron       | 3scale-amp2/backend-rhel7         | d13c45ea48efd2a87cb5efb8b745b80a298cdb789e41995fbdedfcc0bfa6a5c5 | true         | 3scale2.7 | true               | 1.10                 | 1.10             | 3scale2.7    | true               | true                        |             0 |              0 |             0 |
| backend-listener   | 3scale-amp2/backend-rhel7         | d13c45ea48efd2a87cb5efb8b745b80a298cdb789e41995fbdedfcc0bfa6a5c5 | true         | 3scale2.7 | true               | 1.10                 | 1.10             | 3scale2.7    | true               | true                        |             0 |              0 |             0 |
| backend-redis      | rhscl/redis-32-rhel7              | a9bdf52384a222635efc0284db47d12fbde8c3d0fcb66517ba8eefad1d4e9dc9 | true         | 3.2       | true               | 3.2-47               | 3.2-47           | 3.2          | true               | true                        |             0 |              0 |             0 |
| backend-worker     | 3scale-amp2/backend-rhel7         | d13c45ea48efd2a87cb5efb8b745b80a298cdb789e41995fbdedfcc0bfa6a5c5 | true         | 3scale2.7 | true               | 1.10                 | 1.10             | 3scale2.7    | true               | true                        |             0 |              0 |             0 |
| system-app         | 3scale-amp2/system-rhel7          | abe1c074e975b6bf6a1f527a1820671636c393517f2f850f7299c24ae88ba2ab | true         | 3scale2.7 | true               | 3scale2.7.1          | 3scale2.7.1      | 3scale2.7    | true               | true                        |             0 |              0 |             0 |
| system-memcache    | 3scale-amp2/memcached-rhel7       | 9c897d68eebb3a704756b9ddf32358c5b99786055aa1b94cc79be4b9cd3d3f03 | true         | 3scale2.7 | true               | 3scale2.7            | 3scale2.7        | 3scale2.7    | true               | true                        |             0 |              0 |             0 |
| system-mysql       | rhscl/mysql-57-rhel7              | 9a781abe7581cc141e14a7e404ec34125b3e89c008b14f4e7b41e094fd3049fe | true         | 5.7       | true               | 5.7-49               | 5.7-49           | 5.7          | true               | true                        |             0 |              0 |             0 |
| system-redis       | rhscl/redis-32-rhel7              | a9bdf52384a222635efc0284db47d12fbde8c3d0fcb66517ba8eefad1d4e9dc9 | true         | 3.2       | true               | 3.2-47               | 3.2-47           | 3.2          | true               | true                        |             0 |              0 |             0 |
| system-sidekiq     | 3scale-amp2/system-rhel7          | abe1c074e975b6bf6a1f527a1820671636c393517f2f850f7299c24ae88ba2ab | true         | 3scale2.7 | true               | 3scale2.7.1          | 3scale2.7.1      | 3scale2.7    | true               | true                        |             0 |              0 |             0 |
| system-sphinx      | 3scale-amp2/system-rhel7          | abe1c074e975b6bf6a1f527a1820671636c393517f2f850f7299c24ae88ba2ab | true         | 3scale2.7 | true               | 3scale2.7.1          | 3scale2.7.1      | 3scale2.7    | true               | true                        |             0 |              0 |             0 |
| zync               | 3scale-amp2/zync-rhel7            | 90a18fccbc5481a1952841cf6497af3f1e5920310a7c469a26bb2d5b4032ef9f | true         | 3scale2.7 | true               | 3scale2.7.1          | 3scale2.7.1      | 3scale2.7    | true               | true                        |             0 |              0 |             0 |
| zync-database      | rhscl/postgresql-10-rhel7         | de3ab628b403dc5eed986a7f392c34687bddafee7bdfccfd65cecf137ade3dfd | true         | latest    | false              | 1-47                 | 1-47             | latest       | true               | false                       |             0 |              0 |             0 |
| zync-que           | 3scale-amp2/zync-rhel7            | 90a18fccbc5481a1952841cf6497af3f1e5920310a7c469a26bb2d5b4032ef9f | true         | 3scale2.7 | true               | 3scale2.7.1          | 3scale2.7.1      | 3scale2.7    | true               | true                        |             0 |              0 |             0 |
+--------------------+-----------------------------------+------------------------------------------------------------------+--------------+-----------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
+--------------------+-----------------------------------+------------------------------------------------------------------+--------------+-----------+--------------------+----------------------+------------------+--------------+--------------------+-----------------------------+---------------+----------------+---------------+
| COMPONENT          | IMAGE                             | IMAGE HASH                                                       | IMAGE STREAM | TAG       | UPTO DATE WITH TAG | PERSISTENT IMAGE TAG | LATEST PATCH TAG | FLOATING TAG | USING FLOATING TAG | UPTO DATE WITH FLOATING TAG | CRITICAL CVES | IMPORTANT CVES | MODERATE CVES |

```

/lgtm

sergioifg94 committed:
feat(cli): Add `-namespace-pattern` flag

Add optional `-namespace-pattern` flag to CLI in order to filter only namespaces
with names that match the given pattern

sergioifg94 commented:
@Nanyte25 I just pushed a commit to include only namespaces that match a pattern passed as an optional `-namespace-pattern` flag. This pattern is a Go regular expression, so this should do the job for `1.x`:

```sh
./cli -namespace-pattern "openshift-.*"
```

sergioifg94 committed:
fix: Change approach to removing image tag

The `ParseImage` function processes a docker image name to return
a `ClusterImage` struct instance from it. In the process, the previous
approach to remove the tag is to split the image name by ":"  and to
get the first part. This approach fails when there's more than one ":"
in the image, such as when the registry URL has an explicit port. Because
of this, the function would panic with the following error:

```
index out of range [1] with length 1
```

This commit performs the tag separation **after** the image URL
has been split by "/", only to the last part of the resulting
slice. This way the presence of a port won't affect the function.

sergioifg94 commented:
/test e2e

sergioifg94 commented:
/test e2e

Nanyte25 commented:
/lgtm

sergioifg94 commented:
/assign @maleck13

maleck13 commented:
/approve

maleck13 commented:
@Nanyte25 I would suggest not running agianst all namespaces as that will give a lot of results not needed and a lot of additional requests to the registry. I assume the use case is ```redhat-rhmi-*```

Nanyte25 commented:
@maleck13 yes your are correct that was the use case for creating the prefix option instead with`redhat-rhmi-*`
